ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,933 @@
+        clippy::cast_precision_loss
+    )]
+    let proof_length = (u.len() as f64).log(r_f as f64).ceil() as usize + 1usize;
+    // get masks from PRSS
+    let mask_p = ctx.prss().generate_fields(RecordId::from(record)); //(F::ONE, F::ONE);
+    record += 1;
+    let mask_q = ctx.prss().generate_fields(RecordId::from(record));
+    // proof from the prover on the left, i.e. generated via `PRSS.left` rather than receiving it from the left
+    let mut output_left = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        // set mask p(0) from `PRSS.left`
+        mask_p: mask_p.0,
+        // set mask q(0) from `PRSS.left`
+        mask_q: mask_q.0,
+    };
+    // right part of this helpers own proof, to be sent to the left
+    let mut own_proof_right = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        mask_p: F::ZERO,
+        mask_q: F::ZERO,
+    };
+
+    // precomputation for Lagrange
+    let mut e_p = vec![F::ZERO; 2 * r_f];
+    generate_evaluation_points(&mut e_p);
+    // precomputation for interpolation of `u`, `v`, i.e. degree `recursion_factor-1` polynomial
+    let mut ds = vec![F::ONE; r_f];
+    compute_lagrange_denominator(&e_p[0..r_f], &mut ds);
+    let mut base = vec![vec![F::ONE; r_f]; r_f];
+    compute_lagrange_base(&e_p[r_f..2 * r_f], &e_p[0..r_f], &ds, &mut base);
+
+    // iterate over recursions
+    while u.len() >= r_f {
+        // fill u and v with zeros such that it is a multiple of the recursion factor
+        let coset = u.len() % r_f;
+        if coset != 0 {
+            u.extend(std::iter::repeat(F::ZERO).take(coset));
+            v.extend(std::iter::repeat(F::ZERO).take(coset));
+        }
+
+        // amount of polynomials `p`, `q`, i.e. amount of `summand_of_g`
+        let length = u.len() / r_f;
+
+        // define `g` which is represented by `2*recursion_factor` many points.
+        let mut g = vec![F::ZERO; 2 * r_f];
+
+        // compute `g` which is the sum of `summand_of_g`
+        u.chunks(r_f)
+            .zip(v.chunks(r_f))
+            .for_each(|(u, v)| compute_summand_of_g(u, v, r_f, &base, &mut g));
+
+        // compute secret share g_left (i.e. part of the proof recovered by right party) using prss_right
+        let mut output_g_left = Vec::<F>::with_capacity(g.len());
+        // `own_g_left` is needed for hash, i.e. "Fiat-Shamir" `r`:
+        let mut own_g_left = Vec::<F>::with_capacity(g.len());
+        for g in &mut g {
+            // get prss
+            record += 1;
+            let (left, right) = ctx.prss().generate_fields(RecordId::from(record));
+            // push `PRSS.left` to `output_g_left`
+            output_g_left.push(left);
+            // own_g_left:
+            own_g_left.push(right);
+            // compute secret share `own_g_right` by computing `own_g_right=g-own_g_left`
+            // we just need to subtract `own_g_left`, which comes from `PRSS.right`
+            *g -= right;
+        }
```

Comment:
Let's discuss this over a call - I don't really follow right now and I'd like to understand.